### PR TITLE
Use un/tweaked public key types

### PR DESCRIPTION
**Pull request by tcharding**
We have two types for tweaked/untweaked schnorr public keys to help users of the taproot API not mix these two keys up. Currently the `taproot` module uses 'raw' `schnoor::PublicKey`s.

Use the `schnoor` module's tweak/untweaked public key types for the `taproot` API.

Fixes: #725 

Please note, I saw this was labeled 'good-first-issue' but I ignored that and greedily implemented a solution because of two reasons
1. We want to get taproot stuff done post haste.
2. I'm struggling to follow what is going on with all the taproot work so this seemed like a way to get my hands dirty.

### CODE REVIEW

File: src/util/taproot.rs [hunk 1]
```diff
@@ -743,7 +743,7 @@ impl ControlBlock {
         let tweak = TapTweakHash::from_key_and_tweak(self.internal_key, Some(curr_hash));
         self.internal_key.tweak_add_check(
             secp,
-            output_key,
+            &output_key.into_inner(),
```
Comment (sanket1729) [inline src/util/taproot.rs:746]: nit: use as_inner() ?

**Review by sanket1729 — DISMISSED** [review]
ACK 3351927324de518620c0e18f7e52d42bd46a96e8, left a small nit

File: src/util/taproot.rs [hunk 2]
```diff
@@ -268,7 +268,7 @@ impl TaprootSpendInfo {
             internal_key: internal_key,
             merkle_root: merkle_root,
             output_key_parity: parity,
-            output_key: output_key,
+            output_key: TweakedPublicKey::new(output_key),
```
Comment (Kixunil) [inline src/util/taproot.rs:271]: This constructor should've been called `dangerous_assume_tweaked` :(

Comment (sanket1729) [reply src/util/taproot.rs:271]: Yup agreed. I was going to bring this in this review. Don't know how it skipped my review :( during the original one. 

Along with changing the new API to dangerous_assume_tweaked, we can also the logic here to that output_key is computed using the `taptweak` method instead of manually tweaking as implemented here. That way output_key would be of the required type directly.

File: src/util/taproot.rs (hunk 1 above)
Comment (tcharding) [reply src/util/taproot.rs]: Done, squashed, force pushed. Now part of  https://github.com/rust-bitcoin/rust-bitcoin/pull/728/commits/376f654c889f01644e9086e502f08cec0d1b3714

File: src/util/taproot.rs (hunk 2 above)
Comment (tcharding) [reply src/util/taproot.rs]: To implement these suggestions I added 4 additional patches to this PR. The first two of which are cleanup
- Remove comments: https://github.com/rust-bitcoin/rust-bitcoin/pull/728/commits/d84e3590a8436a4229bc7838b12415de9a9dcb74
- Use `debug_assert!` instead of `unreachable!`: https://github.com/rust-bitcoin/rust-bitcoin/pull/728/commits/8424aa13b22bcc78647adf69f7be19eaef1de596
- Implement @Kixunil suggestion, rename constructor: https://github.com/rust-bitcoin/rust-bitcoin/pull/728/commits/1f71b7506836da85c0875f1b0fda510c9ca2c278
- Implement @sanket1729 suggestion, use `tap_tweak` https://github.com/rust-bitcoin/rust-bitcoin/pull/728/commits/6f25d0ce186b48f684c81df9c73c8ba6cc5c95c6 but to do so I return the parity from `tap_tweak` also. This changes the taproot API, please review accordingly.

**Review by sanket1729 — APPROVED** [review]
ACK ec24418cc1bdad98cc14ed094a18397a1a4dc26b. Super clean, no comments :)

File: src/util/schnorr.rs [hunk 3]
```diff
@@ -38,7 +39,10 @@ pub trait TapTweak {
     ///  * H is the hash function
     ///  * c is the commitment data
     ///  * G is the generator point
-    fn tap_tweak<C: Verification>(&self, secp: &Secp256k1<C>, merkle_root: Option<&TapBranchHash>) -> TweakedPublicKey;
+    ///
+    /// # Returns
+    /// The tweaked key and its parity.
+    fn tap_tweak<C: Verification>(&self, secp: &Secp256k1<C>, merkle_root: Option<&TapBranchHash>) -> (TweakedPublicKey, bool);
```
Comment (Kixunil) [inline src/util/schnorr.rs:45]: `bool` is not very clear value for parity. Which is which? I believe it'd be much better to have:

```rust
enum KeyParity {
    Even,
    Odd,
}
```

However, I think this can go to a separate PR given that `parity: bool` was used already.

**Review by Kixunil — DISMISSED** [review]
ACK ec24418cc1bdad98cc14ed094a18397a1a4dc26b

The only issue I have can (and probably should) be addressed in a separate PR.

**Comment by dr-orlovsky** [comment]
Needs rebase :(

File: src/util/schnorr.rs (hunk 3 above)
Comment (tcharding) [reply src/util/schnorr.rs:45]: I'll do the PR, thanks.

Comment (tcharding) [reply src/util/schnorr.rs]: lol, I started doing this over in `rust-secp256k1` and I do not know which is which. Is `true` == `Even` or `Odd`?

Comment (tcharding) [reply src/util/schnorr.rs]: https://github.com/rust-bitcoin/rust-secp256k1/pull/344

File: src/util/schnorr.rs [hunk 4]
```diff
@@ -67,8 +68,9 @@ impl TapTweak for UntweakedPublicKey {
 
 
 impl TweakedPublicKey {
-    /// Create a new [TweakedPublicKey] from a [PublicKey]. No tweak is applied.
-    pub fn new(key: PublicKey) -> TweakedPublicKey {
+    /// Create a new [TweakedPublicKey] from a [PublicKey]. No tweak is applied, consider calling
```
Comment (dr-orlovsky) [inline src/util/schnorr.rs:71]: Nit:
```suggestion
    /// Create a new [`TweakedPublicKey`] from a [`PublicKey`]. No tweak is applied, consider calling
```

File: src/util/schnorr.rs [hunk 5]
```diff
@@ -67,8 +68,9 @@ impl TapTweak for UntweakedPublicKey {
...
+    /// `tap_tweak` on an [UntweakedPublicKey] instead of using this constructor.
```
Comment (dr-orlovsky) [inline src/util/schnorr.rs:72]: Nit:
```suggestion
    /// `tap_tweak` on an [`UntweakedPublicKey`] instead of using this constructor.
```

File: src/util/taproot.rs [hunk 6]
```diff
@@ -250,25 +250,15 @@ impl TaprootSpendInfo {
...
-        let parity = output_key
-            .tweak_add_assign(&secp, &tweak)
-            .expect("TapTweakHash::from_key_and_tweak is broken");
+        let (output_key, parity) = internal_key.tap_tweak(secp, merkle_root);
         Self {
             internal_key: internal_key,
             merkle_root: merkle_root,
             output_key_parity: parity,
-            output_key: output_key,
+            output_key,
```
Comment (dr-orlovsky) [inline src/util/taproot.rs:261]: Nit: with the current code style throughout this repo we have `output_key: output_key` which is set as standard by Andrew

**Review by dr-orlovsky — COMMENTED** [review]
Can you pls squash the commits into one? This is usual requirement for trivial commits by @apoelstra and I do not think that we should diverge from it. Why this is an important? Since we need to test each of the commits to compile and pass all tests, and having commits changing just one line of code makes this really computationally expansive. Another thing is that these all commits introduce _the same change_, so it probably best in terms of history and blame to have these change done in just a one commit.

**Comment by sanket1729** [comment]
Actually, I disagree over here. I like the structure of commits because they are different and do things in an atomic way. _Maybe_ `Add newline to end of file` and `Remove 'what' comments` can go into one `refactor` commit, but apart from that the I think other commits are atomic changes that do one thing. 

Having atomic commits really does aid in my review, and I think it would have been (slightly) more difficult to review if all were in one commit. 

> Since we need to test each of the commits to compile and pass all tests, and having commits changing just one line of code makes this really computationally expansive.

We do want this, but IMO this should never come at the cost of atomic commits. Lines of code have nothing to do with commits, and bitcoin core is full of commits with 10 line descriptions and a few line diffs. 

Infact, I think we should include the text from https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#committing-patches in #587.

**Comment by Kixunil** [comment]
Actually an additional *commit* increases the complexity of bisect using formula `log2(n + 1)` where `n` is number of previous commits. So by very small amount. One would have to double the number of all commits to require one more compilation.

File: src/util/taproot.rs (hunk 6 above)
Comment (tcharding) [reply src/util/taproot.rs:261]: Of course, my mistake. Old habits die hard.

File: src/util/schnorr.rs (hunk 4 above)
Comment (tcharding) [reply src/util/schnorr.rs:71]: Sloppy work by me :), thanks, will add as suggested.

**Comment by tcharding** [comment]
> Can you pls squash the commits into one? This is usual requirement for trivial commits by @apoelstra and I do not think that we should diverge from it. 

I'm happy to argue this one over a beer with @apoelstra but I feel pretty strongly that trivial changes should _not_ be done as part of other work, its jarring for reviewers to have to stop and think 'why is that whitespace added there'. 

> Why this is an important? Since we need to test each of the commits to compile and pass all tests, and having commits changing just one line of code makes this really computationally expansive.

I rekon we should optimize for reviewers not for machine running time on CI, reviewers are expensive machines are cheap.

> Another thing is that these all commits introduce the same change, so it probably best in terms of history and blame to have these change done in just a one commit.

I guess this is subjective to a certain extent, all PRs can be described, at some level, as 'improve xyz'. But like I say above I think it makes review easier if there are many small commits, when folks can trust that my commits don't do strange things its easier for them to catch my strange and silly mistakes.

**Comment by dr-orlovsky** [comment]
utACK b5bf6d73190c7c6a09b20917918ede28baeb5f9c

I just told about my experience in working with this repo for several years: my multi-commit PRs were frequently required to be squashed. I also prefer to have atomic/simple commits.

**Comment by sanket1729** [comment]
I would be really surprised if Andrew is against this. What specifically Andrew mentioned (quite a few times on different PRs from different authors) is one should squash changes in a PR that you have already changed in the same PR.  I think most people(including me) who review commit by commit and find it extra work when something is implemented that is only changed again. I don't mean any of this as any critique of your work, but am just trying to arrive at a consensus on atomic commits.
As an example: https://github.com/rust-bitcoin/rust-bitcoin/pull/499/commits

```
PSBT: adding global types (version, xpub)
PSBT: merging new global keys
PSBT: Key pair serialization for new global keys 
Improving PSBT merge with dedicated 1.29 rustc borrow scope
Nits in new PSBT global types reviews
BSPT: Improving global xpub merging algorithm  
PSBT: Improved global keys version and xpub handling
PSBT global xpub merging algorithm reworked
```
The first three commits are only what is required for the PR. The other commits make changes to the code introduced in the same PR and belong to the previous commits. I understand that sometimes it might be annoying to fix something in a specific commit. On a similar rationale, I had requested to squash only the fixups in the contributing.md PR (that were correctly addressed in the same PR). This is not a hard rule, but based on the complexity of the PR, already existing ACKs from reviewers, subjective decisions can be made to merge fixup commits.

I am sorry for pointing to specific examples, but if you were asked to squash commits it should have been for this reason. If not, it is likely an incorrect recommendation that we should _not_ follow in the future.

**Review by sanket1729 — APPROVED** [review]
ACK b5bf6d73190c7c6a09b20917918ede28baeb5f9c

**Comment by dr-orlovsky** [comment]
@sanket1729 ok, I see, thank you for the explanation.

I do not know why github does not show my ACK above as a review approval. Trying once again...

**Review by dr-orlovsky — APPROVED** [review]
utACK b5bf6d73190c7c6a09b20917918ede28baeb5f9c
